src/Futhark/Analysis/Interference.hs

```diff
@@ -0,0 +1,333 @@
+  [(x, y) | x <- toList xs, y <- toList ys]
+    & foldMap (uncurry f)
+
+analyseStm ::
+  LocalScope KernelsMem m =>
+  LastUseMap ->
+  InUse ->
+  Stm KernelsMem ->
+  m (InUse, LastUsed, Graph VName)
+analyseStm lumap inuse0 stm =
+  inScopeOf stm $ do
+    let pat_name = patElemName $ head $ patternValueElements $ stmPattern stm
+
+    new_mems <-
+      stmPattern stm
+        & patternValueElements
+        & mapM (memInfo . patElemName)
+        <&> catMaybes
+        <&> namesFromList
+
+    -- `new_mems` should interfere with any mems inside the statement expression
+    let inuse_outside = inuse0 <> new_mems
+
+    -- `inuse` is the set of memory blocks that are inuse at the end of any code
+    -- bodies inside the expression. `lus` is the set of all memory blocks that
+    -- have reached their last use in any code bodies inside the
+    -- expression. `graph` is the interference graph computed for any code
+    -- bodies inside the expression.
+    (inuse, lus, graph) <- analyseExp lumap inuse_outside (stmExp stm)
+
+    last_use_mems <-
+      Map.lookup pat_name lumap
+        & fromMaybe mempty
+        & namesToList
+        & mapM memInfo
+        <&> catMaybes
+        <&> namesFromList
+        <&> namesIntersection inuse_outside
+
+    return
+      ( (inuse_outside `namesSubtract` last_use_mems `namesSubtract` lus)
+          <> new_mems,
+        (lus <> last_use_mems) `namesSubtract` new_mems,
+        graph
+          <> cartesian
+            makeEdge
+            (namesToList inuse_outside)
+            (namesToList $ inuse_outside <> inuse <> lus <> last_use_mems)
+      )
+
+analyseExp ::
+  LocalScope KernelsMem m =>
+  LastUseMap ->
+  InUse ->
+  Exp KernelsMem ->
+  m (InUse, LastUsed, Graph VName)
+analyseExp lumap inuse_outside expr =
+  case expr of
+    If _ then_body else_body _ -> do
+      res1 <- analyseBody lumap inuse_outside then_body
+      res2 <- analyseBody lumap inuse_outside else_body
+      return $ res1 <> res2
+    DoLoop _ _ _ body -> do
+      analyseBody lumap inuse_outside body
+    Op (Inner (SegOp segop)) -> do
+      analyseSegOp lumap inuse_outside segop
+    _ ->
+      return mempty
+
+analyseKernelBody ::
+  LocalScope KernelsMem m =>
+  LastUseMap ->
+  InUse ->
+  KernelBody KernelsMem ->
+  m (InUse, LastUsed, Graph VName)
+analyseKernelBody lumap inuse body = analyseStms lumap inuse $ kernelBodyStms body
+
+analyseBody ::
+  LocalScope KernelsMem m =>
+  LastUseMap ->
+  InUse ->
+  Body KernelsMem ->
+  m (InUse, LastUsed, Graph VName)
+analyseBody lumap inuse body = analyseStms lumap inuse $ bodyStms body
+
+analyseStms ::
+  LocalScope KernelsMem m =>
+  LastUseMap ->
+  InUse ->
+  Stms KernelsMem ->
+  m (InUse, LastUsed, Graph VName)
+analyseStms lumap inuse0 stms = do
+  inScopeOf stms $ do
+    foldM
+      ( \(inuse, lus, graph) stm -> do
+          (inuse', lus', graph') <- analyseStm lumap inuse stm
+          return (inuse', lus' <> lus, graph' <> graph)
+      )
```

Comment:
Avoid large lambdas, especially for folds.  Use a `where` clause to define the fold function.  This will also let you write the `foldM` expression itself on a single line.